### DISCUSSION TITLE
How can i get Lighthouse Console output programmatically? Not report, but formatted Lighthouse output.

### DISCUSSION THREAD

**Discussion by JWebDev** [GoogleChrome/lighthouse#14998] [Q&A]
Hi, I would like to have an output that shows the lighthouse in the console. Please note that I launch the lighthouse programmatically, not from the console. If from the console, then I know how to redirect to a file.
`lighthouse https://example.com --output html --output-path ./report.html -GA`

But I just can’t find a method that will return me a formatted output in this style.

```
lighthouse https://example.com --output html --output-path ./report.html -GA
  LH:ChromeLauncher Waiting for browser. +0ms
  LH:ChromeLauncher Waiting for browser... +1ms
  LH:ChromeLauncher Waiting for browser..... +504ms
  LH:ChromeLauncher Waiting for browser.....✓ +3ms
  LH:status Connecting to browser +267ms
  LH:status Navigating to about:blank +20ms
  LH:status Benchmarking machine +22ms
  LH:status Preparing target for navigation mode +1s
  LH:status Navigating to about:blank +15ms
  LH:status Preparing target for navigation +15ms
  LH:status Cleaning origin data +44ms
  LH:status Cleaning browser cache +4ms
  LH:status Preparing network conditions +29ms
  LH:status Navigating to https://example.com/ +50ms
  LH:artifacts:getArtifact DevtoolsLog +3s
  LH:artifacts:getArtifact Trace +0ms
  LH:artifacts:getArtifact DevtoolsLog +2ms
  LH:artifacts:getArtifact Trace +0ms
  LH:artifacts:getArtifact Accessibility +0ms
  LH:artifacts:getArtifact AnchorElements +261ms
  LH:artifacts:getArtifact ConsoleMessages +16ms
  LH:artifacts:getArtifact CSSUsage +0ms
  LH:artifacts:getArtifact Doctype +22ms
  LH:artifacts:getArtifact DOMStats +3ms
  LH:artifacts:getArtifact EmbeddedContent +9ms
  LH:artifacts:getArtifact FontSize +5ms
  LH:artifacts:getArtifact Inputs +13ms
  LH:artifacts:getArtifact GlobalListeners +6ms
  LH:artifacts:getArtifact ImageElements +5ms
  LH:artifacts:getArtifact InstallabilityErrors +12ms
  LH:status Get webapp installability errors +0ms
  LH:artifacts:getArtifact InspectorIssues +1ms
  LH:artifacts:getArtifact JsUsage +1ms
  LH:artifacts:getArtifact LinkElements +0ms
  LH:artifacts:getArtifact MainDocumentContent +6ms
  LH:artifacts:getArtifact MetaElements +5ms
  LH:artifacts:getArtifact NetworkUserAgent +7ms
  LH:artifacts:getArtifact OptimizedImages +0ms
  LH:artifacts:getArtifact ResponseCompression +1ms
  LH:artifacts:getArtifact RobotsTxt +0ms
  LH:artifacts:getArtifact ServiceWorker +254ms
  LH:artifacts:getArtifact Scripts +7ms
  LH:artifacts:getArtifact SourceMaps +0ms
  LH:artifacts:getArtifact Stacks +0ms
  LH:status Collect stacks +0ms
  LH:artifacts:getArtifact TagsBlockingFirstPaint +17ms
  LH:artifacts:getArtifact TapTargets +4ms
  LH:artifacts:getArtifact TraceElements +24ms
  LH:artifacts:getArtifact ViewportDimensions +26ms
  LH:artifacts:getArtifact WebAppManifest +4ms
  LH:status Get webapp manifest +1ms
  LH:artifacts:getArtifact devtoolsLogs +2ms
  LH:artifacts:getArtifact traces +0ms
  LH:artifacts:getArtifact FullPageScreenshot +1ms
  LH:artifacts:getArtifact BFCacheFailures +5s
  LH:status Saving artifacts +257ms
  LH:Artifacts saved to disk in folder: /tmp/test/latest-run +29ms
  LH:status Analyzing and running audits... +1ms
  LH:status Auditing: Uses HTTPS +1ms
  LH:status Auditing: Registers a service worker that controls page and `start_url` +2ms
  LH:status Auditing: Has a `<meta name="viewport">` tag with `width` or `initial-scale` +1ms
  LH:status Auditing: First Contentful Paint +2ms
  LH:status Auditing: Largest Contentful Paint +19ms
  LH:status Auditing: First Meaningful Paint +5ms
  LH:status Auditing: Speed Index +3ms
  LH:status Auditing: Screenshot Thumbnails +205ms
  LH:status Auditing: Final Screenshot +1ms
  LH:status Auditing: Total Blocking Time +1ms
```

-GA options do not make such an output.
I see in the timing array object. But Lighthouse processes format it somehow additionally.

Thanks!

**Comment by adamraine** [comment]
So you want Lighthouse to output the console messages when you run programmatically using our node API? Is my understanding correct?

If that is the case you can use the `logLevel` setting to control how much information Lighthouse outputs. For example, the following code should provide the output like the CLI does:

```js
import lighthouse from 'lighthouse';
lighthouse('https://example.com', {logLevel: 'info'})
```

**Reply by JWebDev** [reply]
> So you want Lighthouse to output the console messages when you run programmatically using our node API? Is my understanding correct?
> 
> If that is the case you can use the `logLevel` setting to control how much information Lighthouse outputs. For example, the following code should provide the output like the CLI does:
> 
> ```js
> import lighthouse from 'lighthouse';
> lighthouse('https://example.com', {logLevel: 'info'})
> ```

Hi @adamraine Thanks for the answer. I will try to explain in other way. 
1. I run Lighthouse programmatic ally with TypeScript. I have some logic and start Lighthouse from my script.
2. Yes I see this information in console, because I am starting the Lighthouse like that
```
const chrome = await launch({chromeFlags: ['--headless'], chromePath: '/usr/bin/google-chrome'})
const flags = {logLevel: 'info', port: chrome.port} as Flags
const runnerResult = await lighthouse('https://example.com', flags)
```
3. I will have this output in variable and save in DB. But not all console output. Only where Chrome starts and ends with time for each operation. Exactly that do Lighthouse output. 
4. How can i generate/get the same output with my `runnerResult: LH.RunnerResult`
5. I hope, that, my explanation is more understandable. Thanks.

**Reply by adamraine** [reply]
The log is not available on `runnerResult`, however you can probably get working solution by using the `lighthouse-logger` package shipped with Lighthouse:

```js
import lighthouse from 'lighthouse';
import log from 'lighthouse-logger';

const events = [];
log.events.addListener('status', e => events.push(e));
lighthouse('https://example.com');
```

Is that closer to what you want?

**Reply by JWebDev** [reply]
@adamraine Thanks, it is very close to what i want. on the events[] i see only two elements status and label. But not `artifacts` and the `time`. Time is more important than artifacts. Can we also get other information, that events will be looking the same as log output? Thanks.

**Reply by adamraine** [reply]
`runnerResult.artifacts.Timing` doesn't give you artifact collection timings but it should :/

I opened https://github.com/GoogleChrome/lighthouse/pull/15024 to hopefully fix that issue.

**Reply by JWebDev** [reply]
@adamraine Thanks for the help. How i get the Timing for statuses? It is possible to get its with the same eventListener?

**Reply by adamraine** [reply]
> It is possible to get its with the same eventListener?

No, that event listener just outputs the console log text. You should use `runnerResult.artifacts.Timing` to get the timing values for artifacts.

**Reply by JWebDev** [reply]
@adamraine Thanks. Understood. And how about timing for statuses? In Console I can see timing for each audit, but in events [] I have only 2 elements ['status', 'status label'].

**Reply by adamraine** [reply]
The timings in the console messages are a part of the [`debug`](https://www.npmjs.com/package/debug) library and are just the time since the last console message. I don't think there is a way to extract them, and they aren't guaranteed to be the same timings we collect in the Lighthouse artifacts.

**Reply by JWebDev** [reply]
@adamraine Understood. And maybe the last question related to my main question.
Maybe there are some possibility from lighthouse, to redirect console output to the file? This would be perfect for me.
